qiskit/extensions/quantum_initializer/state_preparation.py

```diff
@@ -0,0 +1,502 @@
+        self._params_arg = params
+        self._inverse = inverse
+        self._name = "State Preparation Dg" if self._inverse else "State Preparation"
+
+        if isinstance(params, Statevector):
+            params = params.data
+
+        if not isinstance(params, int) and num_qubits is not None:
+            raise QiskitError(
+                "The num_qubits parameter to StatePreparation should only be"
+                " used when params is an integer"
+            )
+        self._from_label = isinstance(params, str)
+        self._from_int = isinstance(params, int)
+
+        num_qubits = self._get_num_qubits(num_qubits, params)
+
+        params = [params] if isinstance(params, int) else params
+
+        super().__init__(self._name, num_qubits, params, label=label)
+
+    def _define(self):
+        if self._from_label:
+            self.definition = self._define_from_label()
+        elif self._from_int:
+            self.definition = self._define_from_int()
+        else:
+            self.definition = self._define_synthesis()
+
+    def _define_from_label(self):
+        q = QuantumRegister(self.num_qubits, "q")
+        initialize_circuit = QuantumCircuit(q, name="init_def")
+
+        for qubit, param in enumerate(reversed(self.params)):
+            if param == "1":
+                initialize_circuit.append(XGate(), [q[qubit]])
+            elif param == "+":
+                initialize_circuit.append(HGate(), [q[qubit]])
+            elif param == "-":
+                initialize_circuit.append(XGate(), [q[qubit]])
+                initialize_circuit.append(HGate(), [q[qubit]])
+            elif param == "r":  # |+i>
+                initialize_circuit.append(HGate(), [q[qubit]])
+                initialize_circuit.append(SGate(), [q[qubit]])
+            elif param == "l":  # |-i>
+                initialize_circuit.append(HGate(), [q[qubit]])
+                initialize_circuit.append(SdgGate(), [q[qubit]])
+
+        if self._inverse:
+            initialize_circuit = initialize_circuit.inverse()
+
+        return initialize_circuit
+
+    def _define_from_int(self):
+        q = QuantumRegister(self.num_qubits, "q")
+        initialize_circuit = QuantumCircuit(q, name="init_def")
+
+        # Convert to int since QuantumCircuit converted to complex
+        # and make a bit string and reverse it
+        intstr = f"{int(np.real(self.params[0])):0{self.num_qubits}b}"[::-1]
+
+        # Raise if number of bits is greater than num_qubits
+        if len(intstr) > self.num_qubits:
+            raise QiskitError(
+                "StatePreparation integer has %s bits, but this exceeds the"
+                " number of qubits in the circuit, %s." % (len(intstr), self.num_qubits)
+            )
+
+        for qubit, bit in enumerate(intstr):
+            if bit == "1":
+                initialize_circuit.append(XGate(), [q[qubit]])
+
+        if self._inverse:
+            initialize_circuit = initialize_circuit.inverse()
+
```

Comment:
This is a special case: As the X gate is it's own inverse, the inverse circuit is the same as the original. So I think we can just save the inversion here 🙂 
```suggestion
    # note: X is it's own inverse, so even if self._inverse is True,
    # we don't need to invert anything
```